pkg/controller/elasticsearch/driver/upscale.go

```diff
@@ -66,33 +69,70 @@ func HandleUpscaleAndSpecChanges(
-			if err != nil {
-				return results, fmt.Errorf("handle volume expansion: %w", err)
-			}
-			if recreateSset {
-				// The StatefulSet is scheduled for recreation: let's requeue before attempting any further spec change.
-				results.Requeue = true
-				continue
-			}
-		}
-		reconciled, err := es_sset.ReconcileStatefulSet(ctx.parentCtx, ctx.k8sClient, ctx.es, res.StatefulSet, ctx.expectations)
+
+	// Check if this is a version upgrade
+	isVersionUpgrade, err := isVersionUpgrade(ctx.es)
+	if err != nil {
+		return results, fmt.Errorf("while checking for version upgrade: %w", err)
+	}
+
+	// If this is not a version upgrade, process all resources normally and return
+	if !isVersionUpgrade {
+		actualStatefulSets, requeue, err := reconcileResources(ctx, actualStatefulSets, adjusted)
 		if err != nil {
-			return results, fmt.Errorf("reconcile StatefulSet: %w", err)
+			return results, fmt.Errorf("while reconciling resources: %w", err)
 		}
-		// update actual with the reconciled ones for next steps to work with up-to-date information
-		actualStatefulSets = actualStatefulSets.WithStatefulSet(reconciled)
+		results.Requeue = requeue
+		results.ActualStatefulSets = actualStatefulSets
+		return results, nil
 	}
+
+	// Version upgrade: separate master and non-master StatefulSets
+	var masterResources, nonMasterResources []nodespec.Resources
+	for _, res := range adjusted {
+		if label.IsMasterNodeSet(res.StatefulSet) {
+			masterResources = append(masterResources, res)
+		} else {
+			nonMasterResources = append(nonMasterResources, res)
+		}
+	}
+
+	// First, reconcile all non-master resources
+	actualStatefulSets, requeue, err := reconcileResources(ctx, actualStatefulSets, nonMasterResources)
+	if err != nil {
+		return results, fmt.Errorf("while reconciling non-master resources: %w", err)
+	}
+	if requeue {
+		results.Requeue = true
+		results.ActualStatefulSets = actualStatefulSets
+		return results, nil
+	}
+
+	targetVersion, err := version.Parse(ctx.es.Spec.Version)
+	if err != nil {
+		return results, fmt.Errorf("while parsing Elasticsearch upgrade target version: %w", err)
+	}
+
+	// Check if all non-master StatefulSets have completed their upgrades before proceeding with master StatefulSets
+	pendingNonMasterSTS, err := findPendingNonMasterStatefulSetUpgrades(ctx.k8sClient, actualStatefulSets, expectedResources.StatefulSets(), targetVersion)
+	if err != nil {
+		return results, fmt.Errorf("while checking non-master upgrade status: %w", err)
+	}
+
+	if len(pendingNonMasterSTS) > 0 {
+		// Non-master StatefulSets are still upgrading, skipping master StatefulSets temporarily.
+		// This will cause a requeue in the caller, and master StatefulSets will attempt to be processed in the next reconciliation
+		ctx.upscaleReporter.RecordPendingNonMasterSTSUpgrades(pendingNonMasterSTS)
+		results.ActualStatefulSets = actualStatefulSets
```

Comment:
Why do these 2 lines depend on `len(pendingNonMasterSTS) > 0`? 
My understanding is that:
* `results.ActualStatefulSets = actualStatefulSets` should be set immediately after the call to `reconcileResources(...)`
* `ctx.upscaleReporter.RecordPendingNonMasterSTSUpgrades(pendingNonMasterSTS)` can be called whether or not `len(pendingNonMasterSTS)` is `0`

(I'm just trying to make this function easier to read/process)